ably/realtime_conn.go

```diff
@@ -87,34 +91,91 @@ func newConn(opts *clientOptions, auth *Auth, callbacks connCallbacks) (*Connect
 	}
 	c.queue = newMsgQueue(c)
 	if !opts.NoConnect {
-		if _, err := c.connect(false); err != nil {
-			return nil, err
-		}
+		go func() {
+			lg := opts.Logger.Sugar()
+			lg.Info("Trying to establish a connection asynchronously")
+			if _, err := c.connect(connArgs{}); err != nil {
+				lg.Errorf("Failed to open connection with err:%v", err)
+			}
+		}()
 	}
-	return c, nil
+	return c
 }
 
-func (c *Connection) dial(proto string, u *url.URL) (proto.Conn, error) {
+func (c *Connection) dial(proto string, u *url.URL) (conn proto.Conn, err error) {
+	lg := c.logger().Sugar()
+	start := time.Now()
+	lg.Debugf("Dial protocol=%q url %q ", proto, u.String())
 	// (RTN23b)
 	query := u.Query()
 	query.Add("heartbeats", "true")
 	u.RawQuery = query.Encode()
+	timeout := c.opts.httpOpenTimeout()
 	if c.opts.Dial != nil {
-		return c.opts.Dial(proto, u)
+		conn, err = c.opts.Dial(proto, u, timeout)
+	} else {
+		conn, err = ablyutil.DialWebsocket(proto, u, timeout)
+	}
+	if err != nil {
+		lg.Debugf("Dial Failed in %v with %v", time.Since(start), err)
+		return nil, err
+	}
+	lg.Debugf("Dial success in %v", time.Since(start))
+	return conn, err
+}
+
+func (c *Connection) connectAfterSuspension(arg connArgs) (Result, error) {
+	timeout := c.opts.suspendedRetryTimeout()
+	lg := c.logger().Sugar()
+	lg.Debugf("Attemting to periodically establish connection after suspension with timeout %v", timeout)
+	tick := time.NewTicker(timeout)
+	for {
+		<-tick.C
+		res, err := c.connectWith(arg)
+		if err != nil {
+			if recoverable(err) {
+				continue
+			}
+			return nil, err
+		}
+		return res, nil
 	}
-	return ablyutil.DialWebsocket(proto, u)
+}
+
+// recoverable returns true if err is recoverable, err is from making a
+// connection
+func recoverable(err error) bool {
+	switch e := err.(type) {
+	case *ErrorInfo:
```

Comment:
Just like we use `errors.Is`, we should use `errors.As` for type-assertions.

Reply:
done 80f6c2a0ef728a62a079f1614dffa99401469987 also addressed https://github.com/ably/ably-go/pull/172#discussion_r512090389 in the same commit